File: packages/core/src/text-buffer-view.ts
import { RGBA } from "./lib/RGBA.js"
import {
  NativeEditorSelectionOperation,
  NativeTextViewCommand,
  NativeError,
  NativeStatus,
  type LineInfo,
  type MeasureResult,
  type RenderLib,
  type ContextTextBufferViewHandle,
} from "./zig.js"
import type { TextBuffer } from "./text-buffer.js"
import type { NativeResourceOwner } from "./buffer.js"
import type { SelectionBehavior, SelectionOccupancy } from "./types.js"

export class TextBufferView {
  private lib: RenderLib
  private native: { scene: NativeResourceOwner; handle: ContextTextBufferViewHandle }
  private textBuffer: TextBuffer
  private _destroyed: boolean = false

  constructor(lib: RenderLib, handle: ContextTextBufferViewHandle, textBuffer: TextBuffer, scene: NativeResourceOwner) {
    if (!scene?.driver) throw new Error("TextBufferView requires an explicit resource owner")
    scene.assertAlive()
    const owner = textBuffer._getOwner()
    if (owner.lib !== lib) throw new Error("TextBufferView library owner mismatch")
    if (owner.scene !== scene || !handle || typeof handle !== "object" || handle.context !== scene.driver.context) {
      throw new Error("TextBufferView Context owner mismatch")
    }
    this.lib = lib
    this.native = { scene, handle }
    this.textBuffer = textBuffer
  }

  static create(textBuffer: TextBuffer): TextBufferView {
    const { lib, scene } = textBuffer._getOwner()
    const handle = lib.createContextTextBufferView(scene.driver.context, textBuffer._getSceneHandle(scene))
    try {
      return new TextBufferView(lib, handle, textBuffer, scene)
    } catch (error) {
      lib.destroyContextTextBufferView(scene.driver.context, handle)
      throw error
    }
  }

  // Fail loud and clear
  private guard(): void {
    if (this._destroyed) throw new Error("TextBufferView is destroyed")
    this.native.scene.assertAlive()
    this.textBuffer._getSceneHandle(this.native.scene)
  }

  /** @internal Drawing targets must use the view's library and Context. */
  public _getOwner(): { lib: RenderLib; scene: NativeResourceOwner } {
    this.guard()
    return { lib: this.lib, scene: this.native.scene }
  }

  /** @internal Requires the exact resource owner. */
  public _getSceneHandle(scene: NativeResourceOwner): ContextTextBufferViewHandle {
    this.guard()
    if (this.native.scene !== scene) throw new Error("TextBufferView Context owner mismatch")
    return this.native.handle
  }

  public setSelection(start: number, end: number, bgColor?: RGBA, fgColor?: RGBA): void {
    this.guard()
    this.lib.contextTextBufferViewSelect(this.native.handle.context, this.native.handle, {
      operation: NativeEditorSelectionOperation.Set,
      start,
      end,
      bg: bgColor,
      fg: fgColor,
    })
  }

  public updateSelection(end: number, bgColor?: RGBA, fgColor?: RGBA): void {
    this.guard()
    this.lib.contextTextBufferViewSelect(this.native.handle.context, this.native.handle, {
      operation: NativeEditorSelectionOperation.Update,
      end,
      bg: bgColor,
      fg: fgColor,
    })
  }

  public resetSelection(): void {
    this.guard()
    this.lib.contextTextBufferViewSelect(this.native.handle.context, this.native.handle, {
      operation: NativeEditorSelectionOperation.Reset,
    })
  }

  public getSelection(): { start: number; end: number } | null {
    this.guard()
    return this.lib.contextTextBufferViewGetInfo(this.native.handle.context, this.native.handle).selection
  }

  public hasSelection(): boolean {
    this.guard()
    return this.getSelection() !== null
  }

  public setLocalSelection(
    anchorX: number,
    anchorY: number,
    focusX: number,
    focusY: number,
    bgColor?: RGBA,
    fgColor?: RGBA,
    behavior: SelectionBehavior = "cell",
  ): boolean {
    this.guard()
    return this.lib.contextTextBufferViewSelect(this.native.handle.context, this.native.handle, {
      operation: NativeEditorSelectionOperation.Local,
      anchorX,
      anchorY,
      focusX,
      focusY,
      bg: bgColor,
      fg: fgColor,
      behavior: behavior === "cell" ? 0 : behavior === "word" ? 1 : 2,
    })
  }

  public updateLocalSelection(
    anchorX: number,
    anchorY: number,
    focusX: number,
    focusY: number,
    bgColor?: RGBA,
    fgColor?: RGBA,
    behavior: SelectionBehavior = "cell",
  ): boolean {
    this.guard()
    return this.lib.contextTextBufferViewSelect(this.native.handle.context, this.native.handle, {
      operation: NativeEditorSelectionOperation.LocalUpdate,
      anchorX,
      anchorY,
      focusX,
      focusY,
      bg: bgColor,
      fg: fgColor,
      behavior: behavior === "cell" ? 0 : behavior === "word" ? 1 : 2,
    })
  }

  public resetLocalSelection(): void {
    this.guard()
    this.lib.contextTextBufferViewSelect(this.native.handle.context, this.native.handle, {
      operation: NativeEditorSelectionOperation.LocalReset,
    })
  }

  public setSelectionOccupancy(occupancy: SelectionOccupancy): void {
    this.guard()
    this.lib.contextTextBufferViewSelect(this.native.handle.context, this.native.handle, {
      operation: NativeEditorSelectionOperation.Occupancy,
      behavior: occupancy === "boundary" ? 1 : 0,
    })
  }

  public getSelectionOccupancy(): SelectionOccupancy {
    this.guard()
    return this.lib.contextTextBufferViewGetInfo(this.native.handle.context, this.native.handle).selectionOccupancy
  }

  public setWrapWidth(width: number | null): void {
    this.guard()
    return this.lib.contextTextBufferViewCommand(
      this.native.handle.context,
      this.native.handle,
      NativeTextViewCommand.WrapWidth,
      width ?? 0,
    )
  }

  public setWrapMode(mode: "none" | "char" | "word"): void {
    this.guard()
    return this.lib.contextTextBufferViewCommand(
      this.native.handle.context,
      this.native.handle,
      NativeTextViewCommand.WrapMode,
      mode === "none" ? 0 : mode === "char" ? 1 : 2,
    )
  }

  public setFirstLineOffset(offset: number): void {
    this.guard()
    return this.lib.contextTextBufferViewCommand(
      this.native.handle.context,
      this.native.handle,
      NativeTextViewCommand.FirstLineOffset,
      offset,
    )
  }

  public setViewportSize(width: number, height: number): void {
    this.guard()
    return this.lib.contextTextBufferViewSetViewport(
      this.native.handle.context,
      this.native.handle,
      { x: 0, y: 0, width, height },
      true,
    )
  }

  public setViewport(x: number, y: number, width: number, height: number): void {
    this.guard()
    return this.lib.contextTextBufferViewSetViewport(this.native.handle.context, this.native.handle, {
      x,
      y,
      width,
      height,
    })
  }

  public get lineInfo(): LineInfo {
    this.guard()
    return this.lib.contextTextBufferViewGetLines(this.native.handle.context, this.native.handle)
  }

  public get logicalLineInfo(): LineInfo {
    this.guard()
    return this.lib.contextTextBufferViewGetLines(this.native.handle.context, this.native.handle, true)
  }

  public getLineSources(startLine: number, lineCount: number): number[] {
    this.guard()
    for (const [value, name] of [
      [startLine, "start line"],
      [lineCount, "line count"],
    ] as const) {
      if (!Number.isInteger(value) || value < 0 || value > 0xffff_ffff) {
        throw new RangeError(`${name} must be a u32 integer`)
      }
    }
    return this.logicalLineInfo.lineSources.slice(startLine, startLine + lineCount)
  }

  public getSelectedText(): string {
    this.guard()
    return this.lib.contextTextBufferViewGetSelectedText(this.native.handle.context, this.native.handle)
  }

  public getPlainText(): string {
    this.guard()
    return this.textBuffer.getPlainText()
  }

  public setTabIndicator(indicator: string | number): void {
    this.guard()
    const codePoint = typeof indicator === "string" ? (indicator.codePointAt(0) ?? 0) : indicator
    return this.lib.contextTextBufferViewCommand(
      this.native.handle.context,
      this.native.handle,
      NativeTextViewCommand.TabIndicator,
      codePoint,
    )
  }

  public setTabIndicatorColor(color: RGBA): void {
    this.guard()
    return this.lib.contextTextBufferViewSetTabColor(this.native.handle.context, this.native.handle, color)
  }

  public setTruncate(truncate: boolean): void {
    this.guard()
    return this.lib.contextTextBufferViewCommand(
      this.native.handle.context,
      this.native.handle,
      NativeTextViewCommand.Truncate,
      truncate ? 1 : 0,
    )
  }

  public measureForDimensions(width: number, height: number): MeasureResult | null {
    this.guard()
    return this.lib.contextTextBufferViewMeasure(this.native.handle.context, this.native.handle, width, height)
  }

  public getVirtualLineCount(): number {
    this.guard()
    return this.lib.contextTextBufferViewGetInfo(this.native.handle.context, this.native.handle).virtualLineCount
  }

  public destroy(): void {
    if (this._destroyed) return
    this.lib.getYogaHost().runMutation(() => {
      if (!this.native.scene.driver.contextDisposed) {
        try {
          this.lib.destroyContextTextBufferView(this.native.handle.context, this.native.handle)
        } catch (error) {
          if (!(error instanceof NativeError) || error.status !== NativeStatus.StaleHandle) throw error
        }
      }
      this._destroyed = true
    })
  }
}

File: packages/core/src/rust/src/text_buffer_view.rs
use crate::ansi::RGBA;
use crate::text_buffer::{TextBuffer, WrapMode};

/// Viewport rectangle.
#[derive(Debug, Clone, Copy)]
pub struct Viewport {
    pub x: u32,
    pub y: u32,
    pub width: u32,
    pub height: u32,
}

/// Cached line info — offsets, widths, source lines, and wrap info.
#[derive(Debug, Clone, Default)]
pub struct LineInfo {
    pub starts: Vec<u32>,
    pub widths: Vec<u32>,
    pub sources: Vec<u32>,
    pub wraps: Vec<u32>,
    pub max_width: u32,
}

/// Selection state.
#[derive(Debug, Clone)]
pub struct Selection {
    pub start: u32,
    pub end: u32,
    pub bg_color: Option<RGBA>,
    pub fg_color: Option<RGBA>,
}

/// Local (visual) selection state.
#[derive(Debug, Clone)]
pub struct LocalSelection {
    pub anchor_x: i32,
    pub anchor_y: i32,
    pub focus_x: i32,
    pub focus_y: i32,
    pub bg_color: Option<RGBA>,
    pub fg_color: Option<RGBA>,
}

/// Measurement result for dimension queries.
#[derive(Debug, Clone, Copy)]
pub struct MeasureResult {
    pub line_count: u32,
    pub max_width: u32,
}

/// A view into a TextBuffer with viewport, wrapping, and selection support.
pub struct TextBufferView {
    tb: *mut TextBuffer,
    _view_id: u32,
    viewport: Viewport,
    wrap_mode: WrapMode,
    wrap_width: Option<u32>,
    selection: Option<Selection>,
    local_selection: Option<LocalSelection>,
    truncate: bool,
    tab_indicator: Option<u32>,
    tab_indicator_color: Option<RGBA>,

    // Cached line info
    cached_line_info: LineInfo,
    cached_logical_line_info: LineInfo,
    cache_epoch: u64,
}

unsafe impl Send for TextBufferView {}

impl TextBufferView {
    /// # Safety
    /// Caller must ensure `tb` lives at least as long as this view.
    pub unsafe fn new(tb: *mut TextBuffer) -> Self {
        let _view_id = (*tb).register_view();
        Self {
            tb,
            _view_id,
            viewport: Viewport {
                x: 0,
                y: 0,
                width: 80,
                height: 24,
            },
            wrap_mode: WrapMode::None,
            wrap_width: None,
            selection: None,
            local_selection: None,
            truncate: false,
            tab_indicator: None,
            tab_indicator_color: None,
            cached_line_info: LineInfo::default(),
            cached_logical_line_info: LineInfo::default(),
            cache_epoch: 0,
        }
    }

    fn tb(&self) -> &TextBuffer {
        unsafe { &*self.tb }
    }

    fn _tb_mut(&mut self) -> &mut TextBuffer {
        unsafe { &mut *self.tb }
    }

    // --- Selection ---

    pub fn set_selection(&mut self, start: u32, end: u32, bg: Option<RGBA>, fg: Option<RGBA>) {
        self.selection = Some(Selection {
            start,
            end,
            bg_color: bg,
            fg_color: fg,
        });
    }

    pub fn reset_selection(&mut self) {
        self.selection = None;
    }

    pub fn pack_selection_info(&self) -> u64 {
        if let Some(ref sel) = self.selection {
            ((sel.start as u64) << 32) | (sel.end as u64)
        } else {
            0
        }
    }

    pub fn update_selection(&mut self, end: u32, bg: Option<RGBA>, fg: Option<RGBA>) {
        if let Some(ref mut sel) = self.selection {
            sel.end = end;
            if bg.is_some() {
                sel.bg_color = bg;
            }
            if fg.is_some() {
                sel.fg_color = fg;
            }
        }
    }

    pub fn set_local_selection(
        &mut self,
        anchor_x: i32,
        anchor_y: i32,
        focus_x: i32,
        focus_y: i32,
        bg: Option<RGBA>,
        fg: Option<RGBA>,
    ) -> bool {
        self.local_selection = Some(LocalSelection {
            anchor_x,
            anchor_y,
            focus_x,
            focus_y,
            bg_color: bg,
            fg_color: fg,
        });
        true
    }

    pub fn update_local_selection(
        &mut self,
        anchor_x: i32,
        anchor_y: i32,
        focus_x: i32,
        focus_y: i32,
        bg: Option<RGBA>,
        fg: Option<RGBA>,
    ) -> bool {
        self.set_local_selection(anchor_x, anchor_y, focus_x, focus_y, bg, fg)
    }

    pub fn reset_local_selection(&mut self) {
        self.local_selection = None;
    }

    /// Get selected text into output buffer.
    pub fn get_selected_text_into_buffer(&self, output: &mut [u8]) -> usize {
        if let Some(ref sel) = self.selection {
            self.tb().get_text_range(sel.start, sel.end, output)
        } else {
            0
        }
    }

    /// Get plain text into output buffer (delegates to TextBuffer).
    pub fn get_plain_text_into_buffer(&self, output: &mut [u8]) -> usize {
        self.tb().get_plain_text_into_buffer(output)
    }

    // --- Viewport ---

    pub fn set_viewport(&mut self, vp: Viewport) {
        self.viewport = vp;
    }

    pub fn set_viewport_size(&mut self, width: u32, height: u32) {
        self.viewport.width = width;
        self.viewport.height = height;
    }

    pub fn set_wrap_width(&mut self, width: Option<u32>) {
        self.wrap_width = width;
    }

    pub fn set_wrap_mode(&mut self, mode: WrapMode) {
        self.wrap_mode = mode;
    }

    pub fn set_truncate(&mut self, truncate: bool) {
        self.truncate = truncate;
    }

    pub fn set_tab_indicator(&mut self, indicator: u32) {
        self.tab_indicator = Some(indicator);
    }

    pub fn set_tab_indicator_color(&mut self, color: RGBA) {
        self.tab_indicator_color = Some(color);
    }

    /// Get the number of virtual lines (after wrapping).
    pub fn get_virtual_line_count(&self) -> u32 {
        // Simplified: no wrapping means virtual == logical
        self.tb().get_line_count()
    }

    /// Get cached line info.
    pub fn get_cached_line_info(&mut self) -> &LineInfo {
        self.rebuild_line_info_if_needed();
        &self.cached_line_info
    }

    /// Get logical (pre-wrap) line info.
    pub fn get_logical_line_info(&mut self) -> &LineInfo {
        self.rebuild_line_info_if_needed();
        &self.cached_logical_line_info
    }

    fn rebuild_line_info_if_needed(&mut self) {
        let epoch = self.tb().content_epoch();
        if self.cache_epoch == epoch {
            return;
        }
        self.cache_epoch = epoch;

        let tb = self.tb();
        let line_count = tb.get_line_count();
        let mut starts = Vec::with_capacity(line_count as usize);
        let mut widths = Vec::with_capacity(line_count as usize);
        let mut sources = Vec::with_capacity(line_count as usize);
        let mut wraps = Vec::with_capacity(line_count as usize);
        let mut max_width: u32 = 0;

        let bytes = tb.rope().to_bytes();
        let mut offset: u32 = 0;
        let mut line: u32 = 0;

        for (i, &b) in bytes.iter().enumerate() {
            if b == b'\n' || i == bytes.len() - 1 {
                let end = if b == b'\n' { i as u32 } else { (i + 1) as u32 };
                let w = end - offset;
                starts.push(offset);
                widths.push(w);
                sources.push(line);
                wraps.push(0);
                if w > max_width {
                    max_width = w;
                }
                offset = (i + 1) as u32;
                line += 1;
            }
        }

        if bytes.is_empty() {
            starts.push(0);
            widths.push(0);
            sources.push(0);
            wraps.push(0);
        }

        self.cached_line_info = LineInfo {
            starts: starts.clone(),
            widths: widths.clone(),
            sources: sources.clone(),
            wraps: wraps.clone(),
            max_width,
        };
        self.cached_logical_line_info = LineInfo {
            starts,
            widths,
            sources,
            wraps,
            max_width,
        };
    }

    /// Measure how many lines and what max width text would need at given dimensions.
    pub fn measure_for_dimensions(&mut self, _width: u32, _height: u32) -> Option<MeasureResult> {
        let line_info = self.get_cached_line_info();
        Some(MeasureResult {
            line_count: line_info.starts.len() as u32,
            max_width: line_info.max_width,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_view_basic() {
        let mut tb = TextBuffer::new(crate::utf8::WidthMethod::Unicode);
        tb.set_text(b"line1\nline2\nline3");
        let view = unsafe { TextBufferView::new(&mut tb) };
        assert_eq!(view.get_virtual_line_count(), 3);
    }

    #[test]
    fn test_view_selection() {
        let mut tb = TextBuffer::new(crate::utf8::WidthMethod::Unicode);
        tb.set_text(b"Hello World");
        let mut view = unsafe { TextBufferView::new(&mut tb) };
        view.set_selection(0, 5, None, None);
        let packed = view.pack_selection_info();
        assert_eq!(packed >> 32, 0);
        assert_eq!(packed & 0xFFFFFFFF, 5);
    }
}
